### PR TITLE
Move Active force rotational diffusion to updater

### LINKED ISSUE #898: Move rotational diffusion to an Updater
## Description

Move the rotational diffusion component of the active force compute into a separate updater.

## Motivation and context

The ActiveForceCompute class modifies particle orientations as part of the force computation. This breaks a number of assumptions in the HOOMD model that force computes only compute forces and do not modify the system state. 

For instance, the forces need to be recomputed when the particle sort is applied. ActiveForceCompute does not do this currently, so it actually may produce incorrect forces on particles on steps where the particle sort occurs.

### PR DESCRIPTION
**Pull request by b-butler**
## Description

This PR moves the rotational diffusion part of `hoomd.md.force.Active` and subclasses to its own updater. In addition this PR addresses various bugs or limitations in other parts of the code that lead to errors or difficulties in adding the updater. For more info on those, see the git commit messages.

## Motivation and context

This addresses #898 and bugs in the existing code base.

Resolves #898

## How has this been tested?

An additional test file has been created for testing the updater.

## Change log

```
Added: A new roational diffusion updater for use with Active Forces ``hoomd.md.update.ActiveRotationalDiffusion``.
```
or
```
Changed: Active force rotaional diffusion is now managed by ``hoomd.md.update.ActiveRotationalDiffusion``.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

**Comment by b-butler** [comment]
This PR needs to be merged after #1053 which shouldn't be a problem since tests won't pass until then.

**Review by joaander — COMMENTED** [review]
The overall structure is fine. I will make a detailed review once tests are passing.

File: hoomd/md/force.py [hunk 1]
```diff
@@ -329,56 +329,66 @@ def update_coeffs(self):  # noqa - will be ported to v3
...
+
+    Attributes:
+        filter (:py:mod:`hoomd.filter`): Subset of particles on which to apply
+            active forces.
+
+    .. py:attribute:: active_force
+
+        active force vector in reference to the orientation of a particle
+        :math:`[\mathrm{force}]`.  It is defined per particle type and stays
+        constant during the simulation.
```
Comment (joaander) [inline hoomd/md/force.py:373]:
```suggestion
        Active force vector in the local reference frame of the particle
        :math:`[\mathrm{force}]`.  It is defined per particle type and stays
        constant during the simulation.
```
Clearer language.

File: hoomd/md/force.py [hunk 2]
```diff
@@ -329,56 +329,66 @@ def update_coeffs(self):  # noqa - will be ported to v3
...
+
+        Type: `TypeParameter` [``particle_type``, `tuple` [`float`, `float`,
+        `float`]]
+
+    .. py:attribute:: active_torque
+
+        active torque vector in reference to the orientation of a particle
+        :math:`[\mathrm{force} \cdot \mathrm{length}]`. It is defined per
+        particle type and stays constant during the simulation.
```
Comment (joaander) [inline hoomd/md/force.py:382]:
```suggestion
        Active torque vector in the local reference frame of the particle
        :math:`[\mathrm{force} \cdot \mathrm{length}]`. It is defined per
        particle type and stays constant during the simulation.
```

File: hoomd/md/force.py [hunk 3]
```diff
@@ -416,32 +426,37 @@ def _attach(self):
...
         super()._attach()
 
+    def create_diffusion_updater(self, trigger, rotational_diffusion):
+        """Create a rotational diffusion updater for this active force.
+
+        Args:
+            trigger (hoomd.trigger.Trigger): Select the timesteps to update
+                rotational diffusion.
+            active_force (hoomd.md.force.Active): The active force associated
+                with the updater. This is not settable after construction.
```
Comment (joaander) [inline hoomd/md/force.py:441]:
```suggestion
```
This is not an argument of the method.

File: hoomd/md/force.py [hunk 4]
```diff
@@ -416,32 +426,37 @@ def _attach(self):
...
-            orientation of a particle :math:`[\mathrm{force} \cdot
-            \mathrm{length}]`. It is defined per particle type and stays
-            constant during the simulation.
+    Args:
+        filter (`hoomd.filter.ParticleFilter`): Subset of particles on which to
+            apply active forces.
+        manifold_constraint (`hoomd.md.manifold.Manifold`): Manifold constraint.
 
     :py:class:`ActiveOnManifold` specifies that a constrained active force
     should be added to all particles similar to :py:class:`Active`. The
```
Comment (joaander) [inline hoomd/md/force.py:462]: "all particles" should instead be "particles selected by the filter."

File: hoomd/md/force.py [hunk 5]
```diff
@@ -464,34 +478,39 @@ class ActiveOnManifold(Force):
...
+    Attributes:
+        filter (`hoomd.filter.ParticleFilter`): Subset of particles on which to
+            apply active forces.
+        manifold_constraint (`hoomd.md.manifold.Manifold`): Manifold constraint.
+
+    .. py:attribute:: active_force
+
+        active force vector in reference to the orientation of a particle
+        :math:`[\mathrm{force}]`.  It is defined per particle type and stays
+        constant during the simulation.
```
Comment (joaander) [inline hoomd/md/force.py:491]:
```suggestion
        Active force vector in the local reference frame of the particle
        :math:`[\mathrm{force}]`.  It is defined per particle type and stays
        constant during the simulation.
```

File: hoomd/md/force.py [hunk 6]
```diff
@@ -464,34 +478,39 @@ class ActiveOnManifold(Force):
...
+
+        Type: `TypeParameter` [``particle_type``, `tuple` [`float`, `float`,
+        `float`]]
+
+    .. py:attribute:: active_torque
+
+        active torque vector in reference to the orientation of a particle
+        :math:`[\mathrm{force} \cdot \mathrm{length}]`. It is defined per
+        particle type and stays constant during the simulation.
```
Comment (joaander) [inline hoomd/md/force.py:500]:
```suggestion
        Active torque vector in local reference frame of the particle
        :math:`[\mathrm{force} \cdot \mathrm{length}]`. It is defined per
        particle type and stays constant during the simulation.
```

File: hoomd/md/methods/methods.py [hunk 7]
```diff
@@ -23,15 +23,7 @@ class Method(_HOOMDBaseObject):
     Note:
         Users should use the subclasses and not instantiate `Method` directly.
     """
-
-    def _detach(self):
```
Comment (joaander) [inline hoomd/md/methods/methods.py:27]: Is this now handled by the base class?

File: hoomd/md/update.py [hunk 8]
```diff
@@ -213,3 +214,95 @@ def _attach(self):
...
+    The diffusion of the updater  follows :math:`\delta \theta / \delta t =
+    \sqrt{2 D_r / \delta t} \Gamma`, where :math:`D_r` is the rotational
+    diffusion constant, and the gamma function is a unit-variance random
+    variable, whose components are uncorrelated in time, space, and between
+    particles. In 3D, :math:`\hat{p}_i` is a unit vector in 3D space, and
+    diffusion follows :math:`\delta \hat{p}_i / \delta t = \sqrt{2 D_r /
+    \delta t} \Gamma (\hat{p}_i (\cos \theta - 1) + \hat{p}_r \sin \theta)`,
+    where :math:`\hat{p}_r` is an uncorrelated random unit vector. The
+    persistence length of an active particle's path is :math:`v_0 / D_r`.
+    The rotational diffusion is applied to the orientation vector/quaternion
```
Comment (joaander) [inline hoomd/md/update.py:236]:
```suggestion
    The rotational diffusion is applied to the orientation quaternion
```

File: hoomd/md/update.py [hunk 9]
```diff
@@ -213,3 +214,95 @@ def _attach(self):
...
+    of each particle. When used with `hoomd.md.force.ActiveOnManifold`,
+    rotational diffusion is performed tangent plane of the manifold.
```
Comment (joaander) [inline hoomd/md/update.py:238]:
```suggestion
    rotational diffusion is performed in the tangent plane of the manifold.
```

File: hoomd/md/update.py [hunk 10]
```diff
@@ -213,3 +214,95 @@ def _attach(self):
...
+
+    Tip:
+        Use `hoomd.md.force.Active.create_diffusion_updater` to construct
+        a `ActiveRotationalDiffusion` instance.
+
+    Args:
+        trigger (hoomd.trigger.Trigger): Select the timesteps to update
+            rotational diffusion.
+        active_force (hoomd.md.force.Active): The active force associated with
```
Comment (joaander) [inline hoomd/md/update.py:247]: Document that `hoomd.md.force.ActiveOnManifold` is also an accepted type.

File: hoomd/md/update.py [hunk 11]
```diff
@@ -213,3 +214,95 @@ def _attach(self):
...
+    persistence length of an active particle's path is :math:`v_0 / D_r`.
+    The rotational diffusion is applied to the orientation vector/quaternion
+    of each particle. When used with `hoomd.md.force.ActiveOnManifold`,
+    rotational diffusion is performed tangent plane of the manifold.
+
+    Tip:
+        Use `hoomd.md.force.Active.create_diffusion_updater` to construct
+        a `ActiveRotationalDiffusion` instance.
+
+    Args:
```
Comment (joaander) [inline hoomd/md/update.py:244]: Move the Args section to the top of the docstring so that it appears close to the argument list when sphinx renders it.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! Here is another round of improvements to the documentation. Also, merge master up into this branch and resolve conflicts.

File: hoomd/md/methods/methods.py (hunk 7 above)
Comment (b-butler) [reply hoomd/md/methods/methods.py:27]: Yes.

**Comment by joaander** [comment]
Some of the tests are failing:
```
FAILED install/hoomd/md/pytest/test_active_rotational_diffusion.py::test_pickling[CPU-Active]
FAILED install/hoomd/md/pytest/test_active_rotational_diffusion.py::test_pickling[CPU-ActiveOnManifold]
```

**Review by joaander — APPROVED** [review]
Thanks!

**Pull request merged**